src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldHaveCorrectSuffix.cs

```diff
@@ -25,24 +29,141 @@ public abstract class IdentifiersShouldHaveCorrectSuffixAnalyzer : DiagnosticAna
+            var interfaceTypeSuffixMap = new Dictionary<INamedTypeSymbol, SuffixInfo>();
+
+            foreach (var tuple in s_baseTypesAndTheirSuffix)
+            {
+                var wellKnownNamedType = context.Compilation.GetTypeByMetadataName(tuple.Item1);
+
+                if (wellKnownNamedType != null && wellKnownNamedType.OriginalDefinition != null)
+                {
+                    // If the type is interface
+                    if (wellKnownNamedType.OriginalDefinition.TypeKind == TypeKind.Interface)
+                    {
+                        interfaceTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                    else
+                    {
+                        baseTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                }
+            }
+
+            if (baseTypeSuffixMap.Count > 0 || interfaceTypeSuffixMap.Count > 0)
+            {
+                context.RegisterSymbolAction((saContext) =>
+                {
+                    var namedTypeSymbol = saContext.Symbol as INamedTypeSymbol;
+                    if (namedTypeSymbol.GetResultantVisibility() != SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    var baseType = namedTypeSymbol.GetBaseTypes().FirstOrDefault(bt => baseTypeSuffixMap.Keys.Contains(bt.OriginalDefinition));
+                    if (baseType != null)
+                    {
+                        var suffixInfo = baseTypeSuffixMap[baseType.OriginalDefinition];
+                        var rule = suffixInfo.CanSuffixBeCollection ? SpecialCollectionRule : DefaultRule;
+                        if ((suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(CollectionString) && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)) ||
+                            (!suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(rule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
+                        }
+
+                        return;
+                    }
+
+                    var implementedInterface = namedTypeSymbol.AllInterfaces.FirstOrDefault(i => interfaceTypeSuffixMap.Keys.Contains(i.OriginalDefinition));
+                    if (implementedInterface != null)
+                    {
+                        var suffixInfo = interfaceTypeSuffixMap[implementedInterface.OriginalDefinition];
+                        if (!namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(DefaultRule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
```

Comment:
@basoundr I am fine with us reporting diagnostics in any order, its not necessary to match FxCop's ordering. I would still be slightly concerned about the later case if we expand the list of these well known types later - I don't know whether or not any real code would implement multiple of these interfaces OR a base type + at least one interface in the list. Anycase, given your code matches the original FXCop implementation, I am okay with retaining your approach :)